worker/runtime/backend_test.go

```diff
@@ -141,9 +143,123 @@ func (s *BackendSuite) TestCreateContainerSetsHandle() {
 	s.NoError(err)
 
 	s.Equal("handle", cont.Handle())
+}
+
+func (s *BackendSuite) TestCreateMaxContainersReached() {
+	backend, err := runtime.NewGardenBackend(s.client,
+		runtime.WithKiller(s.killer),
+		runtime.WithNetwork(s.network),
+		runtime.WithUserNamespace(s.userns),
+		runtime.WithMaxContainers(1),
+		runtime.WithRequestTimeout(1*time.Second),
+	)
+	s.NoError(err)
+
+	fakeTask := new(libcontainerdfakes.FakeTask)
+	fakeContainer := new(libcontainerdfakes.FakeContainer)
+
+	fakeContainer.NewTaskReturns(fakeTask, nil)
+	s.client.NewContainerReturns(fakeContainer, nil)
+
+	s.client.ContainersReturns([]containerd.Container{fakeContainer}, nil)
+	_, err = backend.Create(minimumValidGdnSpec)
+	s.Error(err)
+	s.Contains(err.Error(), "max containers reached")
+}
+
+func (s *BackendSuite) TestCreateMaxContainersReachedConcurrent() {
+	fakeTask := new(libcontainerdfakes.FakeTask)
+	fakeContainer := new(libcontainerdfakes.FakeContainer)
+
+	fakeContainer.NewTaskReturns(fakeTask, nil)
+	s.client.NewContainerReturns(fakeContainer, nil)
+
+	network := new(runtimefakes.FakeNetwork)
+	network.SetupMountsStub = func(string) ([]specs.Mount, error) {
+		s.client.ContainersReturns([]containerd.Container{fakeContainer}, nil)
+		return nil, nil
+	}
+
+	backend, err := runtime.NewGardenBackend(s.client,
+		runtime.WithKiller(s.killer),
+		runtime.WithNetwork(network),
+		runtime.WithUserNamespace(s.userns),
+		runtime.WithMaxContainers(1),
+		runtime.WithRequestTimeout(1*time.Second),
+	)
+	s.NoError(err)
+
+	numberOfRequests := 10
+	requestErrors := make(chan error, numberOfRequests)
+	wg := sync.WaitGroup{}
+	wg.Add(numberOfRequests)
+
+	for i := 0; i < numberOfRequests; i++ {
+		go func() {
+			_, err := backend.Create(minimumValidGdnSpec)
+			if err != nil {
+				requestErrors <- err
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	close(requestErrors)
 
+	s.Len(requestErrors, numberOfRequests-1)
+	s.Equal(s.client.NewContainerCallCount(), 1)
+	for err := range requestErrors {
+		s.Contains(err.Error(), "max containers reached")
+	}
 }
 
+func (s *BackendSuite) TestCreateContainerLockTimeout() {
+	fakeTask := new(libcontainerdfakes.FakeTask)
+	fakeContainer := new(libcontainerdfakes.FakeContainer)
+
+	fakeContainer.IDReturns("handle")
+	fakeContainer.NewTaskReturns(fakeTask, nil)
+	s.client.NewContainerReturns(fakeContainer, nil)
+
+	network := new(runtimefakes.FakeNetwork)
+	network.SetupMountsStub = func(string) ([]specs.Mount, error) {
```

Comment:
It might make more semantic sense to introduce the delay in `s.client.NewContainerReturns(fakeContainer, nil)` which actually creates the container.

Reply:
fixed